### PR TITLE
 feat(tribe-tooltips): Add Tooltip object and views.

### PR DESCRIPTION
**Pull request by Camwyn**
Add a Tooltip object and associated loaders and view files. And styles.

🎫 https://central.tri.be/issues/94274

Needs a TSM II branch to merge into

Required by: https://github.com/moderntribe/event-tickets/pull/1213
and https://github.com/moderntribe/event-tickets-plus/pull/768

### CODE REVIEW

File: src/Tribe/Service_Providers/Tooltip.php [hunk 1]
```diff
@@ -0,0 +1,46 @@
+<?php
+
+/**
+ * Class Tribe__Service_Providers__Tooltip
+ *
+ * @since 4.9
```
Comment (sc0ttkclark) [inline src/Tribe/Service_Providers/Tooltip.php]:
```suggestion
 * @since TBD
```

File: src/Tribe/Tooltips/View.php [hunk 2]
```diff
@@ -0,0 +1,85 @@
+<?php
+
+/**
+ * Class Tribe__Tooltip__View
```
Comment (sc0ttkclark) [inline src/Tribe/Tooltips/View.php]:
```suggestion
 * Class Tribe__Tooltip__View
 *
 * @since TBD
```

File: src/Tribe/Tooltips/View.php [hunk 3]
```diff
@@ -0,0 +1,85 @@
...
+			'classes'   => '',
+			'icon'      => 'info',
+			'direction' => 'down',
+			'context'   => '',
+		];
+
+		$args = wp_parse_args( $original_args, $default_args );
+
+		ob_start();
+		/**
```
Comment (sc0ttkclark) [inline src/Tribe/Tooltips/View.php]:
```suggestion

		/**
```

File: readme.txt [hunk 4]
```diff
@@ -2,6 +2,9 @@
 
 == Changelog ==
 
+= [4.9.4.1] TBD =
+* Tweak - Add new tribe-tooltip object for universal utility [120856]
+
```
Comment (sc0ttkclark) [inline readme.txt]:
```suggestion

= [4.9.5] TBD =


```

File: readme.txt [hunk 5]
```diff
@@ -2,6 +2,9 @@
 
 == Changelog ==
 
+= [4.9.4.1] TBD =
```
Comment (sc0ttkclark) [inline readme.txt]:
```suggestion
= [4.9.6] TBD =

```

Comment (Camwyn) [reply readme.txt]: Yeah, that version number was from when I thought this was just going to be a style tweak.

**Comment by sc0ttkclark** [comment]
@bordoni @lucatume can you have a look at this when you have time?

**Comment by Camwyn** [comment]
This is all working correctly (called in ET/ET+ PRs) and I've fixed the conflicts ⭐️

File: vendor/composer/autoload_real_52.php [hunk 6]
```diff
@@ -2,7 +2,11 @@
 
 // autoload_real_52.php generated by xrstf/composer-php52
 
+<<<<<<< HEAD
```
Comment (bordoni) [inline vendor/composer/autoload_real_52.php]: Conflicts.

File: vendor/composer/autoload_real_52.php [hunk 7]
```diff
@@ -39,6 +49,43 @@ public static function getLoader() {
...
+//		require $vendorDir . '/wp-cli/language-command/language-command.php'; // disabled because of PHP 5.3 syntax
+		require $vendorDir . '/wp-cli/import-command/import-command.php';
+		require $vendorDir . '/wp-cli/extension-command/extension-command.php';
+		require $vendorDir . '/wp-cli/export-command/export-command.php';
+		require $vendorDir . '/wp-cli/eval-command/eval-command.php';
+//		require $vendorDir . '/wp-cli/entity-command/entity-command.php'; // disabled because of PHP 5.3 syntax
+//		require $vendorDir . '/wp-cli/embed-command/embed-command.php'; // disabled because of PHP 5.3 syntax
+		require $vendorDir . '/wp-cli/db-command/db-command.php';
+		require $vendorDir . '/wp-cli/cron-command/cron-command.php';
+		require $vendorDir . '/wp-cli/core-command/core-command.php';
```
Comment (bordoni) [inline vendor/composer/autoload_real_52.php]: Lets make sure WP CLI doesnt get included here.

File: vendor/composer/autoload_psr4.php [hunk 8]
```diff
@@ -6,5 +6,59 @@
...
+    'Webmozart\\Assert\\' => array($vendorDir . '/webmozart/assert/src'),
+    'WP_CLI\\Embeds\\' => array($vendorDir . '/wp-cli/embed-command/src'),
+    'WP_CLI\\AutoloadSplitter\\' => array($vendorDir . '/wp-cli/autoload-splitter/src'),
+    'WP_CLI\\' => array($vendorDir . '/wp-cli/shell-command/src/WP_CLI', $vendorDir . '/wp-cli/entity-command/src/WP_CLI'),
+    'Symfony\\Polyfill\\Mbstring\\' => array($vendorDir . '/symfony/polyfill-mbstring'),
+    'Symfony\\Polyfill\\Ctype\\' => array($vendorDir . '/symfony/polyfill-ctype'),
+    'Symfony\\Component\\Yaml\\' => array($vendorDir . '/symfony/yaml'),
+    'Symfony\\Component\\Translation\\' => array($vendorDir . '/symfony/translation'),
+    'Symfony\\Component\\Process\\' => array($vendorDir . '/symfony/process'),
+    'Symfony\\Component\\Finder\\' => array($vendorDir . '/symfony/finder'),
```
Comment (bordoni) [inline vendor/composer/autoload_psr4.php]: Lets make sure we do  not include the Dev required

File: src/views/tooltip/tooltip.php [hunk 9]
```diff
@@ -0,0 +1,31 @@
...
+ * Tooltip Basic View Template
+ * The base template for Tribe Tooltips.
+ *
+ * Override this template in your own theme by creating a file at [your-theme]/tribe-events/tooltips/tooltip.php
+ *
+ * @package Tribe
+ * @version TBD
+ */
+
+if ( ! defined( 'ABSPATH' ) ) {
```
Comment (bordoni) [inline src/views/tooltip/tooltip.php]: We dont need this.

File: src/Tribe/Tooltip/View.php [hunk 10]
```diff
@@ -0,0 +1,96 @@
...
+			'icon'      => 'info',
+			'direction' => 'down',
+			'context'   => '',
+		];
+
+		$args = wp_parse_args( $original_args, $default_args );
+
+		ob_start();
+
+		$template = $this->get_template_file( 'tooltip' );
```
Comment (bordoni) [inline src/Tribe/Tooltip/View.php]: We can just use `$this->template()` and pass the arguments directly via the context.

File: src/views/tooltip/tooltip.php (hunk 9 above)
Comment (Camwyn) [reply src/views/tooltip/tooltip.php]: Sweet!

File: vendor/composer/autoload_real_52.php (hunk 6 above)
Comment (Camwyn) [reply vendor/composer/autoload_real_52.php]: dang it...why did that not show up for me locally :/

File: vendor/composer/autoload_real_52.php (hunk 7 above)
Comment (Camwyn) [reply vendor/composer/autoload_real_52.php]: Do we want to kill off _all_ the composer entries?

**Comment by Camwyn** [comment]
For the `composer` stuff, I ran `composer install --no-dev` that removed all the changes.

File: readme.txt [hunk 11]
```diff
@@ -2,6 +2,10 @@
 
 == Changelog ==
 
+= [4.9.6] TBD =
+
+* Tweak - Add new tribe-tooltip object for universal utility [120856]
```
Comment (sc0ttkclark) [inline readme.txt]:
```suggestion
* Tweak - Add new `tooltip.view` PHP class to render new tool tips that utilize the `tribe-tooltip` CSS class for universal utility [120856]
```

File: src/Tribe/Tooltip/View.php [hunk 12]
```diff
@@ -0,0 +1,98 @@
...
+		$default_args = [
+			'classes'   => '',
+			'icon'      => 'info',
+			'direction' => 'down',
+			'context'   => '',
+		];
+
+		$args = wp_parse_args( $original_args, $default_args );
+
+		$args[ 'message' ] = $message;
```
Comment (sc0ttkclark) [inline src/Tribe/Tooltip/View.php]:
```suggestion
		// Check for message to be passed.
		if ( empty( $message ) ) {
			return '';
		}

		// Setup message as an array of messages
		$message = (array) $message;
		
		$args['messages'] = $message;
```

File: src/Tribe/Tooltip/View.php [hunk 13]
```diff
@@ -0,0 +1,98 @@
...
+
+		$html = ob_get_clean();
+
+		/**
+		 * Allow us to filter the tooltip output
+		 *
+		 * @since  TBD
+		 *
+		 * @param string $html The tooltip HTML.
+		 * @param array|string $message Array of messages or single message as string.
```
Comment (sc0ttkclark) [inline src/Tribe/Tooltip/View.php]:
```suggestion
		 * @param array $message An array of message strings.
```

File: src/Tribe/Tooltip/View.php [hunk 14]
```diff
@@ -0,0 +1,98 @@
...
+		 * @param array $args Extra arguments, defaults include icon, classes, direction, and context.
+		 */
+		return apply_filters( 'tribe_tooltip_html', $html, $message, $args );
```
Comment (sc0ttkclark) [inline src/Tribe/Tooltip/View.php]:
```suggestion
		return apply_filters( 'tribe_tooltip_html', $html, $messages, $args );
```

**Comment by Camwyn** [comment]
> Needs composer rebuild --no-dev

um... I did `composer install --no-dev` after removing the lock file. As there is no `composer rebuild`

**Pull request merged**